src/themis/secure_cell_seal_passphrase.c

```diff
@@ -0,0 +1,341 @@
+                                                            message_length,
+                                                            user_context,
+                                                            user_context_length,
+                                                            auth_token,
+                                                            auth_token_length,
+                                                            encrypted_message,
+                                                            encrypted_message_length);
+}
+
+themis_status_t themis_auth_sym_decrypt_message_with_passphrase_(const uint8_t* passphrase,
+                                                                 size_t passphrase_length,
+                                                                 const uint8_t* user_context,
+                                                                 size_t user_context_length,
+                                                                 const uint8_t* auth_token,
+                                                                 size_t auth_token_length,
+                                                                 const uint8_t* encrypted_message,
+                                                                 size_t encrypted_message_length,
+                                                                 uint8_t* message,
+                                                                 size_t* message_length)
+{
+    themis_status_t res = THEMIS_FAIL;
+    struct themis_scell_auth_token_passphrase hdr = {0};
+    struct themis_scell_pbkdf2_context kdf = {0};
+    /* Use maximum possible length, not the default one */
+    uint8_t derived_key[THEMIS_AUTH_SYM_MAX_KEY_LENGTH / 8] = {0};
+    size_t derived_key_length = 0;
+
+    res = themis_read_scell_auth_token_passphrase(auth_token, auth_token_length, &hdr);
+    if (res != THEMIS_SUCCESS) {
+        return res;
+    }
+
+    /* Check that message length is consistent with header */
+    if (hdr.message_length != encrypted_message_length) {
+        return THEMIS_FAIL;
+    }
+
+    /*
+     * Verify that algorithm contains a KDF function we support. In particular,
+     * it must not contain SOTER_SYM_NOKDF which is used by master key API.
+     */
+    switch (hdr.alg & SOTER_SYM_KDF_MASK) {
+    case SOTER_SYM_PBKDF2:
+        break;
+    case SOTER_SYM_NOKDF:
+        return THEMIS_FAIL;
+    default:
+        return THEMIS_FAIL;
+    }
+    /* The algorithm also defines length of the derived key we need */
+    derived_key_length = soter_alg_key_length(hdr.alg);
+    /* Algorithm field contains unused bits that must be set to zero */
+    if (!soter_alg_reserved_bits_valid(hdr.alg)) {
+        return THEMIS_FAIL;
+    }
+
+    res = themis_read_scell_pbkdf2_context(&hdr, &kdf);
+    if (res != THEMIS_SUCCESS) {
+        return res;
+    }
+
+    res = soter_pbkdf2_sha256(passphrase,
```

Comment:
same here with hardcoded function. we can use one general, which will return the length of context for specific algo.
then each algorithm has its own function which receives just buffer with the correct length, parse it with own logic (pbkdf2 has 3 fields, some other kdf has other types/lengths of fields) and do what it need to derive key.